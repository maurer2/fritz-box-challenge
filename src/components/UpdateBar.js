import React, { PureComponent } from 'react';
import styled from 'styled-components/macro';
import { PropTypes } from 'prop-types';
import { CSSTransitionGroup } from 'react-transition-group';

const UpdateText = styled.div`
  padding: 1rem;
  text-align: center;
  font-size: 1.5rem;
  font-weight: bold;
  background: red;
`;

const SladeYTransition = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;

  .slide-appear,
  .slide-enter {
    transform: translateY(-100%);
  }

  .slide-appear.slide-appear-active,
  .slide-enter.slide-enter-active {
    transform: translateY(0);
    transition: transform 250ms ease-out;
  }

  .slide-leave {
    transform: translateY(0);
  }

  .slide-leave.slide-leave-active {
    transform: translateY(-100%);
    transition: transform 250ms ease-out;
  }
`;

class UpdateView extends PureComponent {
  render() {
    return (
      <SladeYTransition>
        <CSSTransitionGroup
          component={ React.Fragment }
          transitionAppear={ false }
          transitionName="slide"
          transitionAppearTimeout={ 0 }
          transitionLeaveTimeout={ 250 }
          transitionEnterTimeout={ 250 }
        >
          { this.props.isUpdating && (
            <UpdateText>
              Updating!
            </UpdateText>
          )}
        </CSSTransitionGroup>
      </SladeYTransition>
    );
  }
}

UpdateView.propTypes = {
  isUpdating: PropTypes.bool,
};

export default UpdateView;
