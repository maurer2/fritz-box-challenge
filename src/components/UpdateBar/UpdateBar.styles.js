import styled from 'styled-components/macro';

export const UpdateBar = styled.aside`
  grid-area: updateBar;
  position: relative;
`;

export const SlideYTransition = styled.div`
  > .${(props) => props.transitionName}-appear {
    transform: translateY(-100%);
    transition: transform 500ms ease-out;

    &.${(props) => props.transitionName}-appear-active {
      transform: translateY(0);
    }
  }

  > .${(props) => props.transitionName}-leave {
    transform: translateY(0);
    transition: transform 500ms ease-in;

    &.${(props) => props.transitionName}-leave-active {
      transform: translateY(-100%);
    }
  }
`;

export const Text = styled.div`
  padding: 1rem;
  text-align: center;
  font-size: 1rem;
  background: gray;
`;
