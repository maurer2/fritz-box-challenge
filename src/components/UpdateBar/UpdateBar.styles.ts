import styled, { css } from 'styled-components';

import type { UpdateBarStyleProps } from './UpdateBar.types';

export const UpdateBar = styled.aside<UpdateBarStyleProps>`
  grid-area: updateBar;
  text-align: center;
  background: ${({ theme }) => theme.secondaryColor};
  transform: translateY(0);
  transition: transform 500ms ease-out;
  will-change: transform;
  ${({ $shouldShowUpdateBar }) => !$shouldShowUpdateBar
    && css`
      transform: translateY(-100%);
    `}
`;

export const Text = styled.div`
  padding: 1rem;
  font-size: 1rem;
`;
