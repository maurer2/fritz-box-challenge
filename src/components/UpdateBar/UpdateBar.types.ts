export interface UpdateBarStyleProps {
  isUpdating: boolean;
}
