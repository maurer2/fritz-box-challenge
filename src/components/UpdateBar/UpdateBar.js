import React from 'react';
import { PropTypes } from 'prop-types';
import { CSSTransitionGroup } from 'react-transition-group';

import * as Styles from './UpdateBar.styles';

const Wrapper = ({ children }) => {
  const transitionName = 'slide-vertically';

  return (
    <Styles.UpdateBar>
      <Styles.SlideYTransition transitionName={transitionName}>
        <CSSTransitionGroup
          component={React.Fragment}
          transitionAppear={false}
          transitionEnter={false}
          transitionName={transitionName}
          transitionAppearTimeout={500}
          transitionLeaveTimeout={500}
          transitionEnterTimeout={0}
        >
          { children }
        </CSSTransitionGroup>
      </Styles.SlideYTransition>
    </Styles.UpdateBar>
  );
};

const UpdateBar = ({ isUpdating, isValid }) => {
  const showUpdateBar = (isUpdating || !(isValid));

  return (
    <Wrapper>
      { showUpdateBar && (
        <Styles.Text>
          { !!(isUpdating) && <>Updating!</> }
          { !(isValid) && <>Error!</> }
        </Styles.Text>
      )}
    </Wrapper>
  );
};

const { bool, node } = PropTypes;

UpdateBar.propTypes = {
  isUpdating: bool.isRequired,
  isValid: bool.isRequired,
};

Wrapper.propTypes = {
  children: node,
};

export { UpdateBar };
