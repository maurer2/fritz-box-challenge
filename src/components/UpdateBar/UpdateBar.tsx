import React from 'react';
import PropTypes from 'prop-types';

import { BoxDataContext } from '../DataProvider';

import * as Styles from './UpdateBar.styles';

const UpdateBar: React.FC<{}> = (): JSX.Element => {
  const state = React.useContext(BoxDataContext);

  const isUpdating = 'isUpdating' in state ? state.isUpdating : true;
  const isValid = 'isValid' in state ? state.isValid : false;

  return (
    <Styles.UpdateBar isUpdating={isUpdating}>
      {isUpdating ? (
        <Styles.Text>Updating!</Styles.Text>
      ) : (
        <Styles.Text>
          {!isValid && <>Error!</>}
          {!!isValid && <>Data loaded!</>}
        </Styles.Text>
      )}
    </Styles.UpdateBar>
  );
};

const { bool } = PropTypes;

UpdateBar.propTypes = {
  isUpdating: bool.isRequired,
  isValid: bool.isRequired,
};

export { UpdateBar };
