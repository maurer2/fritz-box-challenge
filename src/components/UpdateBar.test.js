import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import UpdateBar from './UpdateBar';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('UpdateBar', () => {
  const wrapper = shallow(<UpdateBar
    isUpdating={ true }
    isValid={ true }
  />);

  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper.setProps({ isUpdating: false });
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ isUpdating: true });
  });

  it('should have a timer wrapper', () => {
    expect(wrapper.find('Text').length).toBe(1);
  });

  it('should have text in timer wrapper', () => {
    wrapper.setProps({ isUpdating: true });
    expect(wrapper.find('Text').text()).toBe('Updating!');
  });

  it('should have text in timer wrapper', () => {
    wrapper.setProps({ isUpdating: false });
    wrapper.setProps({ isValid: false });

    expect(wrapper.find('Text').text()).toBe('Error!');
  });
});
