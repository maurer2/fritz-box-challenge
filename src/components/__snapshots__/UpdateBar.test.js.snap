// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateBar should match snapshot 1`] = `
<SlideYTransition>
  <CSSTransitionGroup
    component={Symbol(react.fragment)}
    transitionAppear={true}
    transitionAppearTimeout={500}
    transitionEnter={false}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={500}
    transitionName="slide-vertically"
  >
    <Text>
      Updating!
    </Text>
  </CSSTransitionGroup>
</SlideYTransition>
`;

exports[`UpdateBar should match snapshot 2`] = `
<SlideYTransition>
  <CSSTransitionGroup
    component={Symbol(react.fragment)}
    transitionAppear={true}
    transitionAppearTimeout={500}
    transitionEnter={false}
    transitionEnterTimeout={0}
    transitionLeave={true}
    transitionLeaveTimeout={500}
    transitionName="slide-vertically"
  />
</SlideYTransition>
`;
